src/isl/isl_range.rs

```diff
@@ -0,0 +1,850 @@
+            1
+        } else {
+            0
+        };
+        match value.as_i64() {
+            Some(v) => {
+                if v >= min_value {
+                    match v.try_into() {
+                        Err(_) => invalid_schema_error(format!(
+                            "Expected non negative integer greater than {} for range boundary values, found {}",
+                            min_value, v
+                        )),
+                        Ok(non_negative_int_value) => Ok(non_negative_int_value),
+                    }
+                } else {
+                    invalid_schema_error(format!(
+                        "Expected non negative integer greater than {} for range boundary values, found {}",
+                        min_value, v
+                    ))
+                }
+            }
+            None => match value.as_big_int() {
+                None => {
+                    unreachable!("Expected range boundary values must be a non negative integer")
+                }
+                Some(v) => {
+                    if v >= &BigInt::from(min_value) {
+                        match v.try_into() {
+                            Err(_) => invalid_schema_error(format!(
+                                "Expected non negative integer greater than {} for range boundary values, found {}",
+                                min_value,v
+                            )),
+                            Ok(non_negative_int_value) => Ok(non_negative_int_value),
+                        }
+                    } else {
+                        invalid_schema_error(format!(
+                            "Expected non negative integer greater than {} for range boundary values, found {}",
+                            min_value, v
+                        ))
+                    }
+                }
+            },
+        }
+    }
+
+    // helper method to validate range boundary values and construct a `Range`
+    pub(crate) fn validate_and_construct_range(
+        start: TypedRangeBoundaryValue,
+        end: TypedRangeBoundaryValue,
+    ) -> IonSchemaResult<Range> {
+        // validate the range boundary values : `start` and `end`
+        match (&start, &end) {
+            (TypedRangeBoundaryValue::Min, TypedRangeBoundaryValue::Max) => {
+                invalid_schema_error("Range boundaries can not be min and max together (i.e. range::[min, max] is not allowed)")
+            }
+            (TypedRangeBoundaryValue::Max, _) => {
+                invalid_schema_error("Lower range boundary value must not be max")
+            }
+            (_, TypedRangeBoundaryValue::Min) => {
+                invalid_schema_error("Upper range boundary value must not be min")
+            }
+            (_, TypedRangeBoundaryValue::Integer(_)) | (TypedRangeBoundaryValue::Integer(_), _) => {
+                Ok(Range::Integer(RangeImpl::<Integer>::int_range_from_typed_boundary_value(start, end)?))
+            }
```

Comment:
We can eliminate a lot of the boilerplate in this `match` statement if we check that `start.range_type() == end.range_type()` before we run it.

Reply:
The Problem with that would be that `TypedRangeBoundaryValue` has following definition which includes `Min`, `Max`:
```rust
pub(crate) enum TypedRangeBoundaryValue {
    Min,
    Max,
    Integer(RangeBoundaryValue<Integer>),
    IntegerNonNegative(RangeBoundaryValue<usize>),
    TimestampPrecision(RangeBoundaryValue<TimestampPrecision>),
    Float(RangeBoundaryValue<f64>),
    Decimal(RangeBoundaryValue<Decimal>),
    Number(RangeBoundaryValue<Number>),
    Timestamp(RangeBoundaryValue<Timestamp>),
}
```